File: tools/addr2line.py
#!/usr/bin/python
#coding: utf-8

import sys
import re
import subprocess  
import argparse
import struct

class Addr2Line:
    def __init__(self):
        self.__fbase_2_fname = {}
        self.__fbase_2_addrset = {}
        self.__is_shared_object_cache = {}
        self.__address_2_line = {}

    def add(self, address, fbase, fname):
        self.__fbase_2_fname[fbase] = fname
        self.__fbase_2_addrset.setdefault(fbase, set())
        self.__fbase_2_addrset[fbase].add(address)

    def execute(self):
        for fbase, addrset in self.__fbase_2_addrset.items():
            fname = self.__fbase_2_fname[fbase]
            self.__popen(addrset, fbase, fname)

    def get_from_cache(self, address):
        return self.__address_2_line.get(address)

    def __popen(self, addrset, fbase, fname):
        is_so = self.__is_shared_object(fname)

        addrlist = list(addrset)
        if is_so:
            for i in range(len(addrlist)):
                addrlist[i] = addrlist[i] - fbase
        
        pcmd = 'addr2line -Cfse {0}'.format(fname)
        for addr in addrlist:
            pcmd = '{0}  0x{1:x}'.format(pcmd, addr)
        
        p = subprocess.Popen(pcmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        for addr in addrlist:
            addr_raw = addr if not is_so else addr + fbase
            function = self.__to_string(p.stdout.readline()).rstrip('\n')
            file = self.__to_string(p.stdout.readline()).rstrip('\n')
            line = "{0}\t{1}\t{2}".format(function, file, fname)
            self.__address_2_line[addr_raw] = line

    def __is_shared_object(self, file_path):
        if file_path not in self.__is_shared_object_cache: 
            with open(file_path, 'rb') as elf:
                elf.seek(16) # ei_ident
                ei_type = struct.unpack('H', elf.read(2))[0]
                self.__is_shared_object_cache[file_path] = ei_type == 3
        return self.__is_shared_object_cache[file_path]

    def __to_string(self, str_or_bytes): # compatible with Python2 and Python3. 
        if type(str_or_bytes) != str:
            str_or_bytes = str_or_bytes.decode('utf-8')
        return str_or_bytes        

def load_bin_file(file_path):
    with open(file_path, 'rb') as file:
        fmt = '=qqqq28q' # ptr, time, size, reserve, stack
        size = struct.calcsize(fmt)
        data = []
        while True:
            line = file.read(size)
            if not line:
                break
            assert(len(line) == size)
            unpack_data = struct.unpack(fmt, line)
            if unpack_data[0] != 0 and unpack_data[0] != -1:
                for index, value in enumerate(reversed(unpack_data)):
                    if value != 0:
                        index = len(unpack_data) - index
                        break
                line_data = {'ptr' : unpack_data[0], 'time' : unpack_data[1], 'size' : unpack_data[2], 'stack' : unpack_data[4:index]}
                data.append(line_data)
        return data

def save_file(file_path, data):
    with open(file_path, 'w') as file:
        for item in data:
            file.write('time:{0}\tsize:{1}\tptr:0x{2:x}\n'.format(item['time'], item['size'], item['ptr']))
            for frame in item['stack_line']:
                file.write('{0}\n'.format(frame))
            file.write('========\n\n')

def load_data(file_path, maps, output_path):
    data = load_bin_file(file_path)
    data.sort(key = lambda x : x['time'])

    addr2line = Addr2Line()

    for item in data:
        for address in item['stack']:
            line = '0x{:x}'.format(address)
            moudle = list(filter(lambda x : address >= x['start'] and address < x['end'], maps))
            if moudle:
                addr2line.add(address, moudle[0]['start'], moudle[0]['pathname'])

    addr2line.execute()

    for item in data:
        stack_line = []
        for address in item['stack']:
            line = addr2line.get_from_cache(address)
            if not line:
                line = '0x{:x}'.format(address)
            stack_line.append(line)
        item['stack_line'] = stack_line

    save_file(output_path, data)

def _extract_maps(line):
    try:
        maps = re.match(r'([0-9A-Fa-f]+)-([0-9A-Fa-f]+)\s+([-r][-w][-x][sp])\s+([0-9A-Fa-f]+)\s+([:0-9A-Fa-f]+)\s+([0-9A-Fa-f]+)\s+(.*)$', line)
        start = int(maps.group(1), 16)
        end = int(maps.group(2), 16)
        perms = maps.group(3)
        offset = int(maps.group(4), 16)
        dev = maps.group(5)
        inode = maps.group(6)
        pathname = maps.group(7)
    except:
        print('_extract_maps failed:', line)
        raise
    return { 'start': start, 'end': end, 'perms': perms, 'offset': offset, 'dev': dev, 'inode': inode, 'pathname': pathname }

def load_maps(file_path):
    with open(file_path, 'r') as fd:
        maps = map(_extract_maps, fd.readlines())
    data = filter(lambda x : x['perms'] == 'r-xp', maps)
    return list(data)

def analyze(snapshot_path, maps_path, output_path):
    maps = load_maps(maps_path)
    load_data(snapshot_path, maps, output_path)

def _main():
    parser = argparse.ArgumentParser(description='tracing malloc addr2line.')
    parser.add_argument('files', metavar = 'N', type = str, nargs = '+', help='files')
    parser.add_argument('--maps', dest='maps', type = str, help='cache file of /proc/$PID/maps')

    args = parser.parse_args()
    maps = load_maps(args.maps)

    for file_path in args.files:
        load_data(file_path, maps, '{0}.addr2line'.format(file_path))

if __name__ == '__main__':
    _main()

File: tools/addr2line-go/addr2line.go
package main

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"

	"github.com/pkg/errors"
)

var Addr2lineNum int

type Addr2Line struct {
	fbaseToFname    map[uint64]string
	fbaseToAddrset  map[uint64]map[uint64]struct{}
	isSOCache       map[string]bool
	addressToLine   map[uint64]string
	addressToLineMu sync.Mutex
}

func NewAddr2Line() *Addr2Line {
	return &Addr2Line{
		fbaseToFname:   make(map[uint64]string),
		fbaseToAddrset: make(map[uint64]map[uint64]struct{}),
		isSOCache:      make(map[string]bool),
		addressToLine:  make(map[uint64]string),
	}
}

func (a *Addr2Line) Add(address, fbase uint64, fname string) {
	a.fbaseToFname[fbase] = fname
	if _, ok := a.fbaseToAddrset[fbase]; !ok {
		a.fbaseToAddrset[fbase] = make(map[uint64]struct{})
	}
	a.fbaseToAddrset[fbase][address] = struct{}{}
}

func (a *Addr2Line) Execute() error {
	for fbase, addrset := range a.fbaseToAddrset {
		fname := a.fbaseToFname[fbase]

		var addresses []uint64
		for addr := range addrset {
			addresses = append(addresses, addr)
		}
		err := a.addr2line(addresses, fbase, fname)
		if err != nil {
			return err
		}
	}
	return nil
}

func (a *Addr2Line) GetFromCache(address uint64) string {
	return a.addressToLine[address]
}

func (a *Addr2Line) addr2line(addrset []uint64, fbase uint64, fname string) error {
	isSO, err := a.isSharedObject(fname)
	if err != nil {
		return errors.WithStack(err)
	}

	if isSO {
		for i := range addrset {
			addrset[i] = addrset[i] - fbase
		}
	}

	maxAddrLine := len(addrset)/Addr2lineNum + 1

	var wg sync.WaitGroup
	startPos := 0
	for startPos < len(addrset) {
		var addrlist []uint64
		if startPos+maxAddrLine < len(addrset) {
			addrlist = addrset[startPos : startPos+maxAddrLine]
		} else {
			addrlist = addrset[startPos:]
		}
		startPos += maxAddrLine

		wg.Add(1)
		go func() {
			defer wg.Done()

			for i := 0; i < 8; i++ {
				addrlist = a.popen(addrlist, fname, isSO, fbase)
				if addrlist == nil {
					break
				}
			}
		}()
	}

	wg.Wait()
	return nil
}

func (a *Addr2Line) popen(addrlist []uint64, fname string, isSO bool, fbase uint64) []uint64 {
	var stdin bytes.Buffer
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd := exec.Command("addr2line", "-Cfse", fname)
	cmd.Stdin = &stdin
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	for _, v := range addrlist {
		stdin.WriteString(fmt.Sprintf("0x%x\n", v))
	}

	err := cmd.Run()
	if err != nil && stdout.Len() == 0 {
		log.Printf("cmd run failed(0x%x):%v", addrlist[0], err)
		return addrlist
	}

	if stderr.Len() > 0 {
		log.Printf("cmd run error:%s", stderr.String())
	}

	scanner := bufio.NewScanner(&stdout)
	for i, addr := range addrlist {
		addrRaw := addr
		if isSO {
			addrRaw = addr + fbase
		}

		if !scanner.Scan() {
			log.Printf("read function failed:%d, 0x%x", i, addrlist[i])
			return addrlist[i:]
		}
		function := strings.TrimSpace(scanner.Text())

		if !scanner.Scan() {
			log.Printf("read file:line failed:%d, 0x%x", i, addrlist[i])
			return addrlist[i:]
		}
		file := strings.TrimSpace(scanner.Text())

		line := fmt.Sprintf("%s\t%s\t%s", function, file, fname)

		a.addressToLineMu.Lock()
		a.addressToLine[addrRaw] = line
		a.addressToLineMu.Unlock()
	}

	return nil
}

func (a *Addr2Line) isSharedObject(filePath string) (bool, error) {
	if result, ok := a.isSOCache[filePath]; ok {
		return result, nil
	}

	file, err := os.Open(filePath)
	if err != nil {
		return false, errors.WithStack(err)
	}
	defer file.Close()

	// 移动到EI_IDENT位置
	if _, err := file.Seek(16, 0); err != nil {
		return false, errors.WithStack(err)
	}

	var eiType uint16
	if err := binary.Read(file, binary.LittleEndian, &eiType); err != nil {
		return false, errors.WithStack(err)
	}

	isSO := eiType == 3
	a.isSOCache[filePath] = isSO
	return isSO, nil
}

type BinData struct {
	Ptr      int64
	Time     int64
	Size     int64
	Stack    []uint64
	UserData interface{}
}

func loadBinFile(filePath string) ([]BinData, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, errors.WithStack(err)
	}
	defer file.Close()

	data := make([]BinData, 0, 65536)
	buf := make([]byte, 256) // 8*4+8*28 = 256 bytes

	for {
		n, err := file.Read(buf)
		if err != nil || n == 0 {
			break
		}

		if n != len(buf) {
			return nil, errors.Errorf("incorrect file format: %d != %d", n, len(buf))
		}

		reader := bytes.NewReader(buf)

		var item BinData
		var reserve int64

		if err := binary.Read(reader, binary.LittleEndian, &item.Ptr); err != nil {
			return nil, errors.Errorf("read ptr filed: %v", err)
		}
		if err := binary.Read(reader, binary.LittleEndian, &item.Time); err != nil {
			return nil, errors.Errorf("read time filed: %v", err)
		}
		if err := binary.Read(reader, binary.LittleEndian, &item.Size); err != nil {
			return nil, errors.Errorf("read size filed: %v", err)
		}
		if err := binary.Read(reader, binary.LittleEndian, &reserve); err != nil {
			return nil, errors.Errorf("read reserve filed: %v", err)
		}

		item.Stack = make([]uint64, 28)
		if err := binary.Read(reader, binary.LittleEndian, &item.Stack); err != nil {
			return nil, fmt.Errorf("read stack filed: %v", err)
		}

		if item.Ptr <= 0 {
			continue
		}

		lastNonZero := len(item.Stack) - 1
		for ; lastNonZero >= 0 && item.Stack[lastNonZero] == 0; lastNonZero-- {
		}
		if lastNonZero >= 0 {
			item.Stack = item.Stack[:lastNonZero+1]
			data = append(data, item)
		}
	}

	return data, nil
}

type MapInfo struct {
	Start    uint64
	End      uint64
	Perms    string
	Offset   uint64
	Dev      string
	Inode    string
	Pathname string
}

func extractMaps(line string) (MapInfo, error) {
	re := regexp.MustCompile(`([0-9A-Fa-f]+)-([0-9A-Fa-f]+)\s+([-r][-w][-x][sp])\s+([0-9A-Fa-f]+)\s+([:0-9A-Fa-f]+)\s+([0-9A-Fa-f]+)\s+(.*)$`)
	matches := re.FindStringSubmatch(line)

	if len(matches) != 8 {
		return MapInfo{}, fmt.Errorf("read maps failed: %s", line)
	}

	start, err := strconv.ParseUint(matches[1], 16, 64)
	if err != nil {
		return MapInfo{}, fmt.Errorf("read start failed: %v", err)
	}

	end, err := strconv.ParseUint(matches[2], 16, 64)
	if err != nil {
		return MapInfo{}, fmt.Errorf("read end failed: %v", err)
	}

	perms := matches[3]

	offset, err := strconv.ParseUint(matches[4], 16, 64)
	if err != nil {
		return MapInfo{}, fmt.Errorf("read offset failed: %v", err)
	}

	dev := matches[5]
	inode := matches[6]
	pathname := matches[7]

	return MapInfo{
		Start:    start,
		End:      end,
		Perms:    perms,
		Offset:   offset,
		Dev:      dev,
		Inode:    inode,
		Pathname: pathname,
	}, nil
}

func loadMaps(filePath string) ([]MapInfo, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, errors.WithStack(err)
	}
	defer file.Close()

	var maps []MapInfo
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		info, err := extractMaps(line)
		if err != nil {
			log.Printf("warn: %v", err)
			continue
		}

		// 只保留具有r-xp权限的映射
		if info.Perms == "r-xp" {
			maps = append(maps, info)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, errors.WithStack(err)
	}

	return maps, nil
}

func Analyze(snapshotPath string, maps []MapInfo, addr2line *Addr2Line) ([]BinData, error) {
	data, err := loadBinFile(snapshotPath)
	if err != nil {
		return nil, err
	}

	sort.Slice(data, func(i, j int) bool {
		return data[i].Time < data[j].Time
	})

	for _, item := range data {
		for _, address := range item.Stack {
			for _, module := range maps {
				if address >= module.Start && address < module.End {
					addr2line.Add(address, module.Start, module.Pathname)
					break
				}
			}
		}
	}

	err = addr2line.Execute()
	if err != nil {
		return nil, err
	}

	return data, nil
}
